source/portable/NetworkInterface/STM32/NetworkInterface.c

```diff
@@ -0,0 +1,969 @@
+    .ucDuplex = PHY_DUPLEX_AUTO,
+    .ucMDI_X = PHY_MDIX_AUTO,
+};
+
+/*-----------------------------------------------------------*/
+
+void vNetworkInterfaceAllocateRAMToBuffers( NetworkBufferDescriptor_t pxNetworkBuffers[ ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS ] )
+{
+    static uint8_t ucNetworkPackets[ ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS ][ EMAC_TOTAL_BUFFER_SIZE ] __attribute__( ( section( ".EthBuffersSection" ), aligned( 32 ) ) );
+
+    for( size_t ul = 0; ul < ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS; ul++ )
+    {
+        pxNetworkBuffers[ ul ].pucEthernetBuffer = &( ucNetworkPackets[ ul ][ ipBUFFER_PADDING ] );
+        *( ( uint32_t * ) &( ucNetworkPackets[ ul ][ 0 ] ) ) = ( uint32_t ) ( &( pxNetworkBuffers[ ul ] ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static BaseType_t xSTM32_GetPhyLinkStatus( NetworkInterface_t * pxInterface )
+{
+    ( void ) pxInterface;
+
+    BaseType_t xReturn = pdFAIL;
+
+    if( xPhyObject.ulLinkStatusMask != 0U )
+    {
+        xReturn = pdPASS;
+    }
+
+    return xReturn;
+}
+
+/*-----------------------------------------------------------*/
+
+NetworkInterface_t * pxSTM32_FillInterfaceDescriptor( BaseType_t xEMACIndex, NetworkInterface_t * pxInterface )
+{
+    static char pcName[ 17 ];
+
+    snprintf( pcName, sizeof( pcName ), "eth%u", ( unsigned ) xEMACIndex );
+
+    memset( pxInterface, '\0', sizeof( *pxInterface ) );
+    pxInterface->pcName = pcName;
+    pxInterface->pvArgument = ( void * ) xEMACIndex;
+    pxInterface->pfInitialise = xSTM32_NetworkInterfaceInitialise;
+    pxInterface->pfOutput = xSTM32_NetworkInterfaceOutput;
+    pxInterface->pfGetPhyLinkStatus = xSTM32_GetPhyLinkStatus;
+
+    FreeRTOS_AddNetworkInterface( pxInterface );
+
+    return pxInterface;
+}
+
+/*-----------------------------------------------------------*/
+
+static BaseType_t xSTM32_NetworkInterfaceInitialise( NetworkInterface_t * pxInterface )
+{
+    BaseType_t xInitResult = pdFAIL;
+    BaseType_t xResult;
+    HAL_StatusTypeDef xHalResult;
+    NetworkEndPoint_t * pxEndPoint;
+    /* BaseType_t xMACEntry = ETH_MAC_ADDRESS1; */ /* ETH_MAC_ADDRESS0 reserved for the primary MAC-address. */
+
+    static eMAC_INIT_STATUS_TYPE xMacInitStatus = eMACEthInit;
+
+    switch( xMacInitStatus )
+    {
+        default:
+            configASSERT( pdFALSE );
+            break;
+
+        case eMACEthInit:
+            pxMyInterface = pxInterface;
+
+            pxEndPoint = FreeRTOS_FirstEndPoint( pxInterface );
+            configASSERT( pxEndPoint != NULL );
+
+            xEthHandle.Instance = ETH;
+            xEthHandle.Init.MACAddr = ( uint8_t * ) pxEndPoint->xMACAddress.ucBytes; /* FreeRTOS_GetMACAddress(); */
+            xEthHandle.Init.MediaInterface = HAL_ETH_RMII_MODE;
+            xEthHandle.Init.TxDesc = DMATxDscrTab;
+            xEthHandle.Init.RxDesc = DMARxDscrTab;
+            xEthHandle.Init.RxBuffLen = EMAC_DATA_BUFFER_SIZE;
```

Comment:
H7 and H5 drivers require this value to be divisible by 4. Otherwise, HAL_ETH_Init fails. I suggest rounding EMAC_DATA_BUFFER_SIZE up to the nearest multiple of 4.